Sources/BSON/JSON.swift

```diff
@@ -136,4 +136,33 @@ extension JSON {
     }
 }
 
+/// Helpers
+extension JSON {
+    /// Helper function used in `BSONValue` initializers that take in extended JSON and need to
+    /// check that an object has only 1 specified key.
+    ///
+    /// - Parameters:
+    ///   - key: a String representing the one key that the initializer is looking for
+    ///   - `keyPath`: an array of `String`s containing the enclosing JSON keys of the current json being passed in.
+    ///                This is used for error messages.
+    /// - Returns: JSON which is the value at the given `key` in `self`
+    ///            or `nil` if `self` is not an `object` or does not contain the given `key`
+    /// - Throws: `DecodingError` if `self` has too many keys
+    internal func onlyHasKey(key: String, keyPath: [String]) throws -> JSON? {
+        guard case let .object(obj) = self else {
+            return nil
```

Comment:
Not sure what to do in the case that self isn't an object. throw some kind of error? or since in all current usages of this function already check that the json is an object, we can remove this guard?

Reply:
hmm that's a good question.

it seems to me we need to have the guard here right now, even if for no other reason than to extract the `[String: JSON]` as its correct type, right? otherwise you won't be able to do `obj[key]` a few lines down.

maybe this method actually makes more sense as a method to be defined via extension on `Dictionary where Key == String, Value == JSON`. then it would not be possible to call this method on non-objects, as you'd call it on the dictionary you obtain via the switch statements.

alternatively, another option you could consider is to leave this method as-is, and rely on the guard here rather than checking in the individual methods what the type is. that might allow you to remove the switch statements altogether from a number of those implementations where you don't actually use `obj` directly.
it does look like you still use it a fair amount for constructing error messages though, so maybe that wouldn't help much.

Reply:
Oh I like option 3, it would be nice to be able to remove the switch statement entirely for values which only accept object. I can think about ways to construct the error message without `obj` or maybe just leave the whole switch for those ones that use `obj`